### PR TITLE
perf: Implement benchmarking strategy for library core components

### PR DESCRIPTION
**Pull request by gabriel-bornea**
In the `IO` data type, both synchronous and asynchronous error handlers have been refactored to use an optional lifting function, `liftE`, to transform errors. This enhancement allows for better control over error handling. Additionally, the `getOrNull` method and `runAsync` operations have been optimized to create fewer objects, improving performance. A new performance test suite has also been added for better tracking of future modifications.

### CODE REVIEW

File: src/io.ts [hunk 1]
```diff
@@ -208,7 +212,7 @@ export class IO<E, A> {
    * an undefined result. This result is typed as `unknown`, allowing for flexible casting to any
    * desired result type `A` within the `IO` monad's compositional and type-safe framework.
    */
-  static empty = <A>(): IO<unknown, A> => new IO(async () => IO.ok(undefined as any as A));
+  static empty = <A>(): IO<unknown, A> => new IO<unknown, A>(async () => IO.ok(undefined as any as A));
```
Comment (tibtof) [inline src/io.ts:219]: Wouldn't it be better to return IO<never, A> since it's always an A?

File: src/io.ts [hunk 2]
```diff
@@ -45,13 +45,19 @@ export interface Err<E> {
  * @template A The type of the result that the operation may yield upon success.
  */
 export class IO<E, A> {
+  private operations: Array<(input: any) => Promise<Err<E> | Ok<A>>> = [];
```
Comment (tibtof) [inline src/io.ts:48]: I couldn't figure out why this is an array, can there be multiple operations?

Comment (gabriel-bornea) [reply src/io.ts:48]: The `operations` array stores a sequence of asynchronous operations, supporting lazy evaluation where operations are executed only when needed. This method optimizes the execution timing, enhancing overall responsiveness. Additionally, based on performance tests included in this pull request, using this array structure has been shown to reduce memory usage. This reduction results from avoiding the allocation of resources for operations until they are actually required, thus minimizing the memory footprint at runtime.

File: src/io.ts (hunk 1 above)
Comment (gabriel-bornea) [reply src/io.ts]: Yes, you are correct.

**Review by tibtof — APPROVED** [review]
💯

**Pull request merged**